File: src/observers/EventBus.js
/**
 * Observer — подписчики реагируют на события, не зная друг о друге.
 * EventBus = издатель, BudgetObserver = подписчик.
 */
class EventBus {
  constructor() {
    this.listeners = new Map();
  }

  subscribe(eventName, handler) {
    const list = this.listeners.get(eventName) || [];
    list.push(handler);
    this.listeners.set(eventName, list);
  }

  emit(eventName, payload) {
    const list = this.listeners.get(eventName) || [];
    for (const handler of list) {
      handler(payload);
    }
  }
}

module.exports = EventBus;

File: app/src/main/java/com/finmate/app/domain/observer/EventBus.kt
package com.finmate.app.domain.observer

/** Шина событий для Observer. */
class EventBus {
    private val listeners = mutableMapOf<String, MutableList<(Any) -> Unit>>()

    fun subscribe(eventName: String, handler: (Any) -> Unit) {
        listeners.getOrPut(eventName) { mutableListOf() }.add(handler)
    }

    fun emit(eventName: String, payload: Any) {
        listeners[eventName].orEmpty().forEach { it(payload) }
    }
}
